File: 14.longest-common-prefix.cpp
/*
 * @lc app=leetcode id=14 lang=cpp
 *
 * [14] Longest Common Prefix
 */

#include <string>
#include <vector>
#include <assert.h>

using std::vector;
using std::string;

// @lc code=start
class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        if (!strs.size()) {
            return "";
        }
        string ret = strs[0];
        for (const auto& str : strs) {
            size_t size = ret.size();
            for (int i = 0; i < size; i++) {
                if (ret[i] != str[i]) {
                    ret = ret.erase(i);
                    break;
                }
            }
        }
        return ret;
    }
};
// @lc code=end

int
main () {
    Solution s;
    vector<string> strs;
    strs = {};
    assert("" == s.longestCommonPrefix(strs));
    strs = {"flower", "flow", "flight"};
    assert("fl" == s.longestCommonPrefix(strs));
    strs = {"dog", "racecar", "car"};
    assert("" == s.longestCommonPrefix(strs));
    strs = {"a", "ab"};
    assert("a" == s.longestCommonPrefix(strs));
    strs = {"ab", "a"};
    assert("a" == s.longestCommonPrefix(strs));
}

File: 14.longest-common-prefix.rs
/*
 * @lc app=leetcode id=14 lang=rust
 *
 * [14] Longest Common Prefix
 */

use std::cmp::min;

struct Solution;
// @lc code=start
impl Solution {
    pub fn longest_common_prefix(strs: Vec<String>) -> String {
        let mut current = strs[0].len();
        for i in 1..strs.len() {
            current = min(current, strs[i].len());
            for j in 0..current {
                if strs[i - 1].chars().nth(j) != strs[i].chars().nth(j) {
                    current = j;
                    break;
                }
            }
            println!("current = {}", current);
        }
        strs[0].chars().collect::<Vec<char>>()[..current]
            .iter()
            .collect()
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(
        Solution::longest_common_prefix(vec!(
            "flower".to_string(),
            "flow".to_string(),
            "flight".to_string()
        )),
        "fl"
    );
}

#[test]
fn test2() {
    assert_eq!(
        Solution::longest_common_prefix(vec!(
            "dog".to_string(),
            "racecar".to_string(),
            "car".to_string()
        )),
        ""
    );
}
